### PR TITLE
Slurm: Add new option to send USR1 signal before timeout

### LINKED ISSUE #64: Slurm: Send signal if job is about to time out.
Slurm will terminate jobs if they take longer than the requested time.  However, it can be configured to send a signal a bit before that, so the job can save a checkpoint.

For this to work, two changes are necessary:

- Add the argument `--signal=USR1@120` to the sbatch script.  This will tell Slurm to send a SIGUSR1 (a user-defined signal) at least 120 seconds before the timeout.  Might make sense to make the 120 configurable, for jobs where saving takes longer.
- Execute the user script with `srun` (i.e. `srun python ...` instead of directly `python ...`) in the sbatch script.  Otherwise the signal will not be send to the Python process.

The user then still needs to set up signal handler in their code to detect it and take appropriate actions (e.g. save a checkpoint) but since this is very task-specific, it is up to the user to take care of it.  A simple example should be added to the documentation, though.

For reference, here is a working example which doesn't use cluster_utils:

sbatch script:
```bash
#!/bin/bash

#SBATCH --partition=cpu-galvani
#SBATCH --time=00:03:00  # job will be cancelled after 3 minutes
#SBATCH --signal=USR1@30  # signal about 30 seconds before timeout
#SBATCH --output=output-%j.out

echo "Start execution"
srun python3.11 ./signal_listening_job.py
rc=$?

echo "signal_listening_job.py terminated with exit code ${rc}"
```

signal_listening_job.py:
```python
"""A simple example script that does some dummy work and listens for a stop signal."""
import signal
import sys
import time


class Worker:

    N_STEPS = 100
    STEP_DURATION_S = 10

    def __init__(self) -> None:
        self.step = 0
        self.received_signal = False

        # register signal handler
        signal.signal(signal.SIGUSR1, self.signal_handler_sigusr1)

    def signal_handler_sigusr1(self, sig, frame):
        """This method gets called when the process receives the SIGUSR1 signal."""
        print("Received SIGUSR1, prepare for shutdown.", flush=True)
        self.received_signal = True

    def run(self) -> None:
        """Do the actual work here."""
        while (self.step < self.N_STEPS) and not self.received_signal:
            print("Step {}".format(self.step), flush=True)
            # do some dummy work (aka sleep)
            time.sleep(self.STEP_DURATION_S)

            self.step += 1

        if self.received_signal:
            print("Save current status and exit")
            time.sleep(5)
            ...
            # Nothing saved in this example but in an actual application this might be
            # the place to store a snapshot.

            # Simply return with a custom error code in this example.  When using
            # cluster utils, this would be the place to call `exit_for_resume()`.
            return 3

        return 0


if __name__ == "__main__":
    sys.exit(Worker().run())
```

### PR DESCRIPTION
**Pull request by luator**
# TL;DR
Add option to send signal to jobs before they are killed by timeout.

# Long Version
Add a new option `timeout_signal_time` in the cluster_requirements section.  If set, it adds `--signal=USR1@<time>` to the `sbatch` call, resulting in Slurm to send a USR1 signal at the specified time (in seconds) before it would kill the job due to timeout.

For the signal to actually be sent to the process running the main script, this needs to be executed with `srun`, so some change in the job command creation is needed to inject that for Slurm jobs.

Also add an example and embed it in the documentation.

Resolves #64.

# How I Tested
By running the example.

### CODE REVIEW

**Review by mseitzer — COMMENTED** [review]
Very nice, thanks! Some comments:

1) I would like to have a more consistent handling how time is configured. Right now, we have two entries with time, `request_time` and `timeout_signal_time`, where one is a duration string, and one is an integer. This is confusing to the user. Possible resolutions that come to my mind (in order of work):

- Rename `timeout_signal_time` to `timeout_signal_time_in_secs`. Cumbersome to write, but helps the user.
- Specify `timeout_signal_time` as a duration string. Requires us to parse it into seconds internally.
- Allow both types (seconds as int, duration as string) for all time entries, storing them into a single format internally and then converting as required. Probably most convenient for the user.

2) Is the signaling working with containers as well? Is singularity handling the signal itself, or is it passing it on to the running command?

3) I was thinking if we should extend the client API by doing the signal handling part internally, i.e. the user can then just check something like `cluster_utils.should_exit_for_resume()` in their training code, and react accordingly. (This does not need to be part of this PR.)

**Comment by luator** [comment]
1. I think adding `..._in_secs` is a good idea to make the format/unit clear.  I'm generally not so happy with the name `timeout_signal_time`, as I think it's not very descriptive.  Something like `seconds_before_timeout_at_which_signal_is_sent` would be much clearer but also annoyingly long...  
  I wouldn't unifying the format with `request_time` as the parsing done by Slurm for that is actually not so trivial (from the sbatch manpage: 'Acceptable time formats include "minutes", "minutes:seconds", "hours:minutes:seconds", "days-hours", "days-hours:minutes" and "days-hours:minutes:seconds".'), so I'd rather not deal with that in our code.
3. Yes, I tested with Singularity because I was also a bit worried about that.  It was working fine, so seems the signal is properly passed on by Singularity.  From what I understood, it might not work directly if you have a bash script wrapping your main script, though (but I didn't explicitly test that).
4. Sounds good to me.  It will make it a bit less flexible but much easier for the user.  I'd indeed rather do that in a separate PR, though.

**Comment by luator** [comment]
How about `signal_seconds_before_timeout` for the parameter name?

File: cluster_utils/job.py [hunk 1]
```diff
@@ -84,7 +84,19 @@ def generate_final_setting(self, paths):
...
         return current_setting
 
-    def generate_execution_cmd(self, paths):
+    def generate_execution_cmd(self, paths, cmd_prefix: Optional[str] = None):
+        """Generate the commands to execute the job.
+
+        Args:
+            paths: Dictionary containing relevant paths.
+            cmd_prefix: String that is prepended to the command that runs the user
+                script.  Can be used to wrap in in some cluster-system-specific command
```
[the remark this answers is not in the document]
Comment (mseitzer) [reply cluster_utils/job.py:93]: Hm. Do you use it consistently? With things like this, I am fine with it as long as it's consistent :)

**Comment by mseitzer** [comment]
> 2\. From what I understood, it might not work directly if you have a bash script wrapping your main script, though (but I didn't explicitly test that).

Maybe noting this in the docs would be nice then.

File: cluster_utils/job.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (luator) [reply cluster_utils/job.py:93]: Yes, I think I'm consistent with it (it's pretty much in my muscle memory by now).  But of course it will likely not be consistent with parts written by other people.

**Comment by luator** [comment]
> Maybe noting this in the docs would be nice then.

I'll actually do an explicit test first, so we know for sure.

**Review by mseitzer — APPROVED** [review]
LGTM

**Comment by luator** [comment]
Wrapper scripts are indeed a problem for the signal.  I couldn't directly find a proper way of handling it in this case, so I just added a warning to the docs for now.
